### PR TITLE
Add reset/passwd capability for local users

### PR DESCRIPTION
**Pull request by alex-kovoy**
This PR contains the following changes:

1. Replaces SignupToken and SignupTokenData with resources `UserToken` and `UserTokenSecrets`.
2. Implements `tctl users reset <username>` command.
3. Adds new AudtEvent to capture UserToken creation. 

Note, this PR is against `alexey/webapps` branch.

Fixes https://github.com/gravitational/teleport/issues/3265

### CODE REVIEW

**Comment by alex-kovoy** [comment]
Once this PR reviewed, I will send another with `E` changes.

File: lib/auth/user.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
+/*
+Copyright 2015 Gravitational, Inc.
```
Comment (alex-kovoy) [inline lib/auth/user.go]: FYI, `new_web_user.go` has been renamed to `user.go`. No code changes in regards to `user` functions.

**Review by fspmarshall — COMMENTED** [review]
None of the APIs being removed affect compatibility between `teleport` instances of different versions, right?  Just UI and `tctl` stuff?

**Comment by alex-kovoy** [comment]
> None of the APIs being removed affect compatibility between `teleport` instances of different versions, right? Just UI and `tctl` stuff?

Only `proxy` and `auth` should be upgraded to the same version. Given upcoming UI redesign changes, I do not see a way how we can keep 2 UI APIs versions compatible. It does not impact node instances.

**Comment by benarent** [comment]
RE: 
> Given upcoming UI redesign changes, I do not see a way how we can keep 2 UI APIs versions compatible. It does not impact node instances.

This falls into the world of https://github.com/gravitational/teleport/issues/3278 and I'm a little anxious of breaking compatibility. I chatted with @russjones and if we do, we need to plan accordingly.  I'm going to write up risks and benefits of this approach into a doc and setup a meeting to review with the team.

**Comment by alex-kovoy** [comment]
On a second thought, most of the old web ui will work fine with the new `auth` version. Only the signup flow will be broken.

File: lib/auth/apiserver.go [hunk 2]
```diff
@@ -103,9 +102,11 @@ func NewAPIServer(config *APIConfig) http.Handler {
 	srv.POST("/:version/users/:user/ssh/authenticate", srv.withAuth(srv.authenticateSSHUser))
 	srv.GET("/:version/users/:user/web/sessions/:sid", srv.withAuth(srv.getWebSession))
 	srv.DELETE("/:version/users/:user/web/sessions/:sid", srv.withAuth(srv.deleteWebSession))
-	srv.GET("/:version/signuptokens/:token", srv.withAuth(srv.getSignupTokenData))
```
Comment (russjones) [inline lib/auth/apiserver.go]: I'm concerned about these. Why can't these be maintained for older proxies and but marked with `// DELETE IN: 4.4 (or 5.0)` so they can be removed in the next version of Teleport?

I understand you've done some refactoring around the internals, but can't the handler transform the old request to the new format?

File: lib/auth/apiserver.go [hunk 3]
```diff
@@ -1123,6 +1123,53 @@ func (s *APIServer) getClusterCACert(auth ClientI, w http.ResponseWriter, r *htt
 	return localCA, nil
 }
 
+func (s *APIServer) rotateUserTokenSecrets(auth ClientI, w http.ResponseWriter, r *http.Request, p httprouter.Params, version string) (interface{}, error) {
```
Comment (russjones) [inline lib/auth/apiserver.go]: All new APIs should be GRPC not HTTP/JSON which has been deprecated. Take a look at https://github.com/gravitational/teleport/blob/master/lib/auth/grpcserver.go.

File: lib/auth/auth_with_roles.go [hunk 4]
```diff
@@ -1041,41 +1034,38 @@ func (a *AuthWithRoles) GenerateUserCerts(ctx context.Context, req proto.UserCer
...
-	return a.authServer.GetSignupTokenData(token)
+	return a.authServer.CreateSignupU2FRegisterRequest(token)
 }
 
-func (a *AuthWithRoles) GetSignupToken(token string) (*services.SignupToken, error) {
-	// signup token are their own authz resource
-	return a.authServer.GetSignupToken(token)
-}
+func (a *AuthWithRoles) CreateUserToken(req CreateUserTokenRequest) (services.UserToken, error) {
+	if err := a.action(defaults.Namespace, services.KindUser, services.VerbUpdate); err != nil {
```
Comment (russjones) [inline lib/auth/auth_with_roles.go]: Why was this changed from `services.VerbCreate` to `services.VerbUpdate`?

File: lib/auth/auth_with_roles.go [hunk 5]
```diff
@@ -1041,41 +1034,38 @@ func (a *AuthWithRoles) GenerateUserCerts(ctx context.Context, req proto.UserCer
...
 	// tokens are their own authz mechanism, no need to double check
-	return a.authServer.CreateUserWithoutOTP(token, password)
+	return a.authServer.RotateUserTokenSecrets(tokenID)
 }
 
-func (a *AuthWithRoles) CreateUserWithU2FToken(token string, password string, u2fRegisterResponse u2f.RegisterResponse) (services.WebSession, error) {
-	// signup tokens are their own authz resource
-	return a.authServer.CreateUserWithU2FToken(token, password, u2fRegisterResponse)
+func (a *AuthWithRoles) ChangePasswordWithToken(req ChangePasswordWithTokenRequest) (services.WebSession, error) {
+	// token is it's own auth, no need for extra auth
```
Comment (russjones) [inline lib/auth/auth_with_roles.go]: Comments should be full sentences, so start with a capital letter and end with a period. You don't need to change old news, but new ones should follow this format.

File: lib/auth/password.go [hunk 6]
```diff
@@ -178,23 +257,110 @@ func (s *AuthServer) getOTPType(user string) (string, error) {
...
+	switch cap.GetSecondFactor() {
+	case teleport.OFF:
+		return nil
+	case teleport.OTP, teleport.TOTP, teleport.HOTP:
+		secrets, err := s.Identity.GetUserTokenSecrets(req.TokenID)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		// TODO: create a separate method to validate TOTP without inserting it first
```
Comment (russjones) [inline lib/auth/password.go]: Is this TODO still valid, if not, please remove the comment.

File: lib/auth/password.go [hunk 7]
```diff
@@ -178,23 +257,110 @@ func (s *AuthServer) getOTPType(user string) (string, error) {
...
+	err = s.changeUserSecondFactor(req, userToken)
 	if err != nil {
-		return "", nil, trace.Wrap(err)
+		return nil, trace.Wrap(err)
 	}
 
-	// create otp url
-	params := map[string][]byte{"secret": []byte(otpSecret)}
-	otpURL := utils.GenerateOTPURL("totp", user, params)
+	err = s.deleteUserTokens(username)
```
Comment (russjones) [inline lib/auth/password.go]: Shouldn't the users second factor and password be updated, then the token deleted?

File: lib/auth/password.go [hunk 8]
```diff
@@ -178,23 +257,110 @@ func (s *AuthServer) getOTPType(user string) (string, error) {
 	return teleport.HOTP, nil
 }
 
-// GetOTPData returns the OTP Key, Key URL, and the QR code.
-func (s *AuthServer) GetOTPData(user string) (string, []byte, error) {
-	// get otp key from backend
-	otpSecret, err := s.GetTOTP(user)
+func (s *AuthServer) changePasswordWithToken(req ChangePasswordWithTokenRequest) (services.User, error) {
```
Comment (russjones) [inline lib/auth/password.go]: Can group some of the logic and add comments within? Makes it easier to follow. Something like:

```go
func (s *AuthServer) changePasswordWithToken(req ChangePasswordWithTokenRequest) (services.User, error) {
	// Get cluster configuration and check if local auth is allowed.
	clusterConfig, err := s.GetClusterConfig()
	if err != nil {
		return nil, trace.Wrap(err)
	}
	if clusterConfig.GetLocalAuth() == false {
		return nil, trace.AccessDenied(noLocalAuth)
	}

	// Check if the signup token exists.
	userToken, err := s.GetUserToken(req.TokenID)
	if err != nil {
		return nil, trace.Wrap(err)
	}
	if userToken.Expiry().Before(s.clock.Now().UTC()) {
		return nil, trace.BadParameter("expired token")
	}

	// Check the password meets the complexity requirements and update it.
	err = services.VerifyPassword(req.Password)
	if err != nil {
		return nil, trace.Wrap(err)
	}
	err = s.UpsertPassword(username, []byte(req.Password))
	if err != nil {
		return nil, trace.Wrap(err)
	}

	[...]
}
```

File: lib/auth/password.go [hunk 9]
```diff
@@ -10,13 +10,61 @@ import (
...
+
+	sess, err := s.createUserWebSession(user)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return sess, nil
+}
+
+// ResetPassword securely generates a new random password and assigns it to user
```
Comment (russjones) [inline lib/auth/password.go]: The comment should explain that this function is used to invalidate a user during the password reset process.

File: lib/services/usertoken.go [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+	// GetURL returns URL
+	GetURL() string
+	// SetURL returns URL
+	SetURL(string)
+	// CheckAndSetDefaults checks and set default values for any missing fields.
+	CheckAndSetDefaults() error
+}
+
+// UserTokenV3 is an user token format V3
+type UserTokenV3 struct {
```
Comment (russjones) [inline lib/services/usertoken.go]: This should be defined in `types.proto` same for `UserTokenSpecV3`.

File: lib/services/usertokensecrets.go [hunk 11]
```diff
@@ -0,0 +1,205 @@
...
+	// GetQRCode returns QRCode
+	GetQRCode() []byte
+	// GetOTPKey returns OTP key
+	GetOTPKey() string
+	// CheckAndSetDefaults checks and set default values for any missing fields.
+	CheckAndSetDefaults() error
+}
+
+// UserTokenSecretsV3 is an user token spec format V3
+type UserTokenSecretsV3 struct {
```
Comment (russjones) [inline lib/services/usertokensecrets.go]: This should be defined in `types.proto` same for `UserTokenSecretsSpecV3 `.

File: lib/web/apiserver.go [hunk 12]
```diff
@@ -1177,32 +1163,76 @@ func (h *Handler) renewSession(w http.ResponseWriter, r *http.Request, _ httprou
...
+	return ui.UserToken{
+		URL: userToken.GetURL(),
+		Expiry: userToken.Expiry(),
+	}, nil
+}
+
+func (h *Handler) getUserTokenHandle(w http.ResponseWriter, r *http.Request, p httprouter.Params) (interface{}, error) {
+	result, err := h.getUserToken(p[0].Value)
+	if err != nil {
+		log.Warnf("failed to fetch user token: %v", err)
```
Comment (russjones) [inline lib/web/apiserver.go]: Log lines should begin with capital letters.

File: lib/web/apiserver.go [hunk 13]
```diff
@@ -1177,32 +1163,76 @@ func (h *Handler) renewSession(w http.ResponseWriter, r *http.Request, _ httprou
...
-		User:        user,
-		QR:          qrCodeBytes,
+	return ui.UserToken{
+		URL: userToken.GetURL(),
+		Expiry: userToken.Expiry(),
+	}, nil
+}
+
+func (h *Handler) getUserTokenHandle(w http.ResponseWriter, r *http.Request, p httprouter.Params) (interface{}, error) {
+	result, err := h.getUserToken(p[0].Value)
```
Comment (russjones) [inline lib/web/apiserver.go]: Check length of `p` before accessing by index.

File: tool/tctl/common/user_command.go [hunk 14]
```diff
@@ -111,35 +151,66 @@ func (u *UserCommand) Add(client auth.ClientI) error {
...
+	}
+
+	traits := map[string][]string{
+		teleport.TraitLogins:     strings.Split(u.allowedLogins, ","),
+		teleport.TraitKubeGroups: kubeGroups,
+	}
+
+	user.SetTraits(traits)
+	user.AddRole(teleport.AdminRoleName)
+	err = client.UpsertUser(user)
```
Comment (russjones) [inline tool/tctl/common/user_command.go]: Why isn't all of this logic within `CreateUserToken`?

File: lib/auth/auth_with_roles.go (hunk 4 above)
Comment (fspmarshall) [reply lib/auth/auth_with_roles.go]: This strikes me as a naming issue; I think `CreateUserToken` is creating a *password reset* token, which does effectively map to `VerbUpdate` with our current model.

File: lib/auth/apiserver.go (hunk 2 above)
Comment (alex-kovoy) [reply lib/auth/apiserver.go]: Old functionality requires most of the existing endpoints as the flow uses below methods as well.
```
CreateUserWithoutOTP
CreateUserWithOTP
CreateUserWithU2FToken
etc
```

My thinking was it's take too much time to ensure backward compatibility between old and new  signup UI flows. And since these are sensitive (security) changes I think it might be good to keep them to minimal.

File: lib/auth/auth_with_roles.go (hunk 4 above)
Comment (alex-kovoy) [reply lib/auth/auth_with_roles.go]: We no longer create a user when processing signup token. `UserToken` is about resetting a password and generating a secure token to set up a new one. Which is essentially user update operation.

File: lib/auth/password.go (hunk 6 above)
Comment (alex-kovoy) [reply lib/auth/password.go]: Yes, it is valid.

File: lib/auth/password.go (hunk 7 above)
Comment (alex-kovoy) [reply lib/auth/password.go]: Since these are not part of one atomic operation, it's better to delete a token first before issuing DB writes as we do not want to end up with a valid token + partially updated user.

File: tool/tctl/common/user_command.go (hunk 14 above)
Comment (alex-kovoy) [reply tool/tctl/common/user_command.go]: CreateUserToken is about resetting a password. In here,  `tctl users add xxx` creates a new user and generates a token to set up a password.

File: lib/auth/auth_with_roles.go (hunk 4 above)
Comment (alex-kovoy) [reply lib/auth/auth_with_roles.go]: Does it make sense to rename `CreateUserToken` to `CreatePasswordResetToken` and consequanlty `UserToken` to `PasswordResetToken` ? The former gives a little bit room for other use cases moving forward as far as naming is concern.

File: lib/auth/password.go (hunk 8 above)
Comment (alex-kovoy) [reply lib/auth/password.go]: Currently it has the flow which minimizes the chances of partially updated user + valid token. I will add more comments.

File: lib/auth/auth_with_roles.go (hunk 4 above)
Comment (alex-kovoy) [reply lib/auth/auth_with_roles.go]: Renamed to ResetPasswordToken

File: lib/auth/grpcserver.go [hunk 15]
```diff
@@ -246,6 +247,84 @@ func (g *GRPCServer) SetAccessRequestState(ctx context.Context, req *proto.Reque
 	return &empty.Empty{}, nil
 }
 
+func (g *GRPCServer) CreateResetPasswordToken(ctx context.Context, req *proto.CreateResetPasswordTokenRequest) (*services.ResetPasswordTokenV3, error) {
+	auth, err := g.authenticate(ctx)
+	if err != nil {
+		return nil, trail.ToGRPC(err)
+	}
+
+	if req == nil {
```
Comment (russjones) [inline lib/auth/grpcserver.go]: Why are you creating it if it's `nil`? Wont `Name` and `Type` be empty below?

File: lib/auth/grpcserver.go [hunk 16]
```diff
@@ -246,6 +247,84 @@ func (g *GRPCServer) SetAccessRequestState(ctx context.Context, req *proto.Reque
...
+		req = &proto.CreateResetPasswordTokenRequest{}
+	}
+
+	ResetPasswordToken, err := auth.CreateResetPasswordToken(ctx, CreateResetPasswordTokenRequest{
```
Comment (russjones) [inline lib/auth/grpcserver.go]: Lowercase variable name.

File: lib/auth/grpcserver.go [hunk 17]
```diff
@@ -246,6 +247,84 @@ func (g *GRPCServer) SetAccessRequestState(ctx context.Context, req *proto.Reque
...
+	if err != nil {
+		return nil, trail.ToGRPC(err)
+	}
+
+	tokenID := ""
+	if req != nil {
+		tokenID = req.TokenID
+	}
+
+	ResetPasswordToken, err := auth.GetResetPasswordToken(ctx, tokenID)
```
Comment (russjones) [inline lib/auth/grpcserver.go]: Lowercase variable name.

File: lib/services/resetpasswordtoken.go [hunk 18]
```diff
@@ -0,0 +1,184 @@
...
+			"type": ["string"]
+		},
+		"url": {
+			"type": ["string"]
+		}
+  }
+}`
+
+// UnmarshalResetPasswordToken unmarshals ResetPasswordToken
+func UnmarshalResetPasswordToken(bytes []byte) (ResetPasswordToken, error) {
```
Comment (russjones) [inline lib/services/resetpasswordtoken.go]: Do you know why we have a package level lock and interface for all other resources? Wondering if by not doing this here we will run into a problem later.

**Review by russjones — APPROVED** [review]

File: lib/auth/auth_with_roles.go [hunk 19]
```diff
@@ -1041,41 +1034,38 @@ func (a *AuthWithRoles) GenerateUserCerts(ctx context.Context, req proto.UserCer
...
 	// signup token are their own authz resource
-	return a.authServer.GetSignupTokenData(token)
+	return a.authServer.CreateSignupU2FRegisterRequest(token)
 }
 
-func (a *AuthWithRoles) GetSignupToken(token string) (*services.SignupToken, error) {
-	// signup token are their own authz resource
-	return a.authServer.GetSignupToken(token)
-}
+func (a *AuthWithRoles) CreateResetPasswordToken(ctx context.Context, req CreateResetPasswordTokenRequest) (services.ResetPasswordToken, error) {
```
Comment (russjones) [inline lib/auth/auth_with_roles.go]: @alex-kovoy This should also be limited so a user can only reset their own password or `tctl` can reset any password.

Comment (alex-kovoy) [reply lib/auth/auth_with_roles.go]: Right now only admin (who has `services.KindUser, services.VerbUpdate` permissions) can do that. Users without this permission are not allowed to reset passwords even their own. 

There is a separate API which allows a user to change own password (Account screen) but this is not the same as reset. 

Given that, what should be done differently in this regard?

Comment (alex-kovoy) [reply lib/auth/auth_with_roles.go]: + there is a new UI coming which allows admins to reset user passwords from WEB UI.

Comment (russjones) [reply lib/auth/auth_with_roles.go]: Okay, makes sense.
